### PR TITLE
Upgrade Motor Types and Imports

### LINKED ISSUE #555: Refactor Motor to bring in a lot of the things that ended up in targets
Bring Keychain into motor

### PR DESCRIPTION
**Pull request by prdlk**
## Changes

* Refactored Types
* Cleaned up some methods
* Updated some paths

### New Features *(required)*

Added additional methods for managing `x/registry` in the binded motor node.

* `BuyAlias()`
* `TransferAlias()`
* `SellAlias()`

### Deprecations *(required)*

This PR serves to clean up the `third_party` proto imports for motor. By modifying the generated proto-go package to follow standard naming convention the following import paths will change.

<table><tbody><tr><th colspan="1" rowspan="1"><p><strong>Previous Import Path</strong></p></th><th colspan="1" rowspan="1"><p><strong>New Import Path</strong></p></th></tr><tr><td colspan="1" rowspan="1"><p><code class="proseCode">github.com/sonr-io/sonr/third_party/types/service</code></p></td><td colspan="1" rowspan="1"><p><code class="proseCode">github.com/sonr-io/sonr/third_party/types/motor/api/v1/service/v1</code></p></td></tr><tr><td colspan="1" rowspan="1"><p><code class="proseCode">github.com/sonr-io/sonr/third_party/types/motor</code></p></td><td colspan="1" rowspan="1"><p><code class="proseCode">github.com/sonr-io/sonr/third_party/types/motor/api/v1</code></p></td></tr></tbody></table>

### Enhancements *(optional)*

Added `SendTx(routeUrl string, msg sdk.Msg) ([]byte, error)` to `pkg/motor` as a generic method to send transactions through the binded motor node. This method is being utilized by the new Alias management methods

## Checklist

* [ ] Unit tests

* [ X ] Documentation

## References *(optional)*

Fixes
Connects

 Mention [stepsize] in a comment if you'd like to report some technical debt. See examples here.

### CODE REVIEW

File: bind/motor-mobile/bind.go [hunk 1]
```diff
@@ -214,22 +220,55 @@ func IssuePayment(buf []byte) ([]byte, error) {
...
 	resp := mt.StatResponse{
 		Address:     instance.GetAddress(),
 		Balance:     int32(instance.GetBalance()),
-		DidDocument: diddoc,
+		DidDocument: didDoc,
 	}
 	return resp.Marshal()
 }
+
+func BuyAlias(buf []byte) ([]byte, error) {
```
Comment (joshLong145) [inline bind/motor-mobile/bind.go:236]: `Buy Alias` should be implemented within the motor pkg and called from this wrapper. reduces duplication for given targets and make adding support for speedway easier.

File: bind/motor-mobile/bind.go [hunk 2]
```diff
@@ -214,22 +220,55 @@ func IssuePayment(buf []byte) ([]byte, error) {
...
+		return nil, ct.ErrMotorWallet
+	}
+	var request rt.MsgBuyAlias
+	if err := request.Unmarshal(buf); err != nil {
+		return nil, fmt.Errorf("unmarshal request: %s", err)
+	}
+	return instance.SendTx(request.Route(), &request)
+}
+
+func SellAlias(buf []byte) ([]byte, error) {
```
Comment (joshLong145) [inline bind/motor-mobile/bind.go:254]: `Sell Alias` should be implemented within the motor pkg and called from this wrapper. reduces duplication for given targets and make adding support for speedway easier.

File: bind/motor-mobile/bind.go [hunk 3]
```diff
@@ -214,22 +220,55 @@ func IssuePayment(buf []byte) ([]byte, error) {
...
+		return nil, ct.ErrMotorWallet
+	}
+	var request rt.MsgSellAlias
+	if err := request.Unmarshal(buf); err != nil {
+		return nil, fmt.Errorf("unmarshal request: %s", err)
+	}
+	return instance.SendTx(request.Route(), &request)
+}
+
+func TransferAlias(buf []byte) ([]byte, error) {
```
Comment (joshLong145) [inline bind/motor-mobile/bind.go:272]: `Transfer Alias` should be implemented within the motor pkg and called from this wrapper. reduces duplication for given targets and make adding support for speedway easier.

File: proto/motor/v1/request.proto [hunk 4]
```diff
@@ -73,7 +73,7 @@ message QueryRequest {
 message PaymentRequest {
   string to = 1;
   string from = 2;
-  int64 amount = 3;
+  int32 amount = 3;
```
Comment (ntindle) [inline proto/motor/v1/request.proto]: ?

File: proto/motor/v1/service.proto [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+// Package Motor is used for defining a Motor node and its properties.
+package sonrio.motor.api.v1;
+
+import "motor/v1/request.proto";
+import "motor/v1/response.proto";
+
+option go_package = "github.com/sonr-io/sonr/third_party/types/motor/api/v1";
+
+service MotorService {
```
Comment (joshLong145) [inline proto/motor/v1/service.proto]: i think we should remove this new rpc service interface for now. code could be moved to a new branch to revisit.

File: proto/motor/v1/response.proto [hunk 6]
```diff
@@ -21,8 +21,8 @@ message InitializeResponse {
 
 message CreateAccountResponse {
   bytes aes_psk = 1;
-  string address = 2;
-  bytes did_document = 3;
+  bytes aes_dsc = 2;
```
Comment (joshLong145) [inline proto/motor/v1/response.proto]: is it okay to plain text transport this key? I think this might be a security risk.

**Review by joshLong145 — CHANGES_REQUESTED** [review]

File: bind/motor-mobile/bind.go [hunk 7]
```diff
@@ -1,28 +1,25 @@
...
-	callback MotorCallback
+	objectBuilders map[string]*object.ObjectBuilder
+	instance       mtr.MotorNode
+	callback       MotorCallback
 )
 
 type MotorCallback interface {
 	OnDiscover(data []byte)
-	OnWalletCreated(ok bool)
+	OnWalletEvent(msg string, isDone bool)
```
Comment (mcjcloud) [inline bind/motor-mobile/bind.go]: I think having a method for each event makes sense. It's more verbose and readable.

File: bind/motor-mobile/bind.go [hunk 8]
```diff
@@ -33,30 +30,39 @@ func Init(buf []byte, cb MotorCallback) ([]byte, error) {
...
-		return resp.Marshal()
 	}
-	return nil, errors.New("loading existing account not implemented")
+	return resp.Marshal()
 }
 
 func CreateAccount(buf []byte) ([]byte, error) {
 	if instance == nil {
-		return nil, errWalletNotExists
+		return nil, ct.ErrMotorWallet
```
Comment (mcjcloud) [inline bind/motor-mobile/bind.go]: The error name here, "MotorWallet", isn't very descriptive

File: pkg/motor/create_account.go [hunk 9]
```diff
@@ -126,7 +113,7 @@ func createWhoIs(m *motorNodeImpl) (*sdk.TxResponse, error) {
 	if err != nil {
 		return nil, err
 	}
-
+	msg1.Type()
```
Comment (mcjcloud) [inline pkg/motor/create_account.go]: What is this for?

File: pkg/motor/create_account.go [hunk 10]
```diff
@@ -179,9 +166,29 @@ func pskEncrypt(shard []byte) ([]byte, []byte, error) {
...
 }
 
-func dscEncrypt(shard, dsc []byte) ([]byte, error) {
-	if len(dsc) != 32 {
-		return nil, errors.New("dsc must be 32 bytes")
+// dscEncrypt encrypts the shard with the DSC key
+// Returns: encrypted shard, given key, error
+func dscEncrypt(shard, d []byte) ([]byte, []byte, error) {
+	// Check if the DSC is valid
+	if len(d) != 32 {
```
Comment (mcjcloud) [inline pkg/motor/create_account.go]: This creates unexpected behavior. If the caller provides a key, they'll expect that to be used.

File: pkg/motor/motor.go [hunk 11]
```diff
@@ -244,3 +246,18 @@ func (w *motorNodeImpl) AddCredentialVerificationMethod(id string, cred *did.Cre
 
 	return nil
 }
+
+func (w *motorNodeImpl) SendTx(routeUrl string, msg sdk.Msg) ([]byte, error) {
```
Comment (mcjcloud) [inline pkg/motor/motor.go:250]: This allows arbitrary transactions to be sent from motor. Do we want that?

File: proto/common/v1/info.proto [hunk 12]
```diff
@@ -37,3 +37,13 @@ enum BlockchainModule {
...
 }
+
+// AuthInfo is a object used by Motor clients to store authentication information in Biometric storage
+message AuthInfo {
+    string address = 1;
+    string did = 2;
+    bytes aes_dsc_key = 3;
+    bytes aes_psk_key = 4;
+    string password = 5;
+    int64 timestamp = 6;
```
Comment (mcjcloud) [inline proto/common/v1/info.proto:48]: What is the timestamp for?

File: proto/motor/v1/request.proto (hunk 4 above)
Comment (mcjcloud) [inline proto/motor/v1/request.proto]: You're using `Int64` in code, but `int32` here.

File: proto/motor/v1/request.proto [hunk 13]
```diff
@@ -143,3 +143,6 @@ message SeachBucketContentBySchemaRequest {
   // schema identifier
   string schema_did = 3;
 }
+
+message StatRequest {
```
Comment (mcjcloud) [inline proto/motor/v1/request.proto:150]: Is this necessary?

**Review by mcjcloud — CHANGES_REQUESTED** [review]

File: proto/motor/v1/response.proto (hunk 6 above)
Comment (mcjcloud) [reply proto/motor/v1/response.proto]: +1

My refactor for keychain in motor will remove this from the request/response models anyway.

File: bind/motor-mobile/bind.go (hunk 7 above)
Comment (joshLong145) [reply bind/motor-mobile/bind.go]: + 1

File: bind/motor-mobile/bind.go (hunk 1 above)
Comment (ntindle) [reply bind/motor-mobile/bind.go:236]: Shouldn't provide focus on only speedway. Should be holistic for all platforms, and keep in mind embeded targets

File: bind/motor-mobile/bind.go (hunk 2 above)
Comment (ntindle) [reply bind/motor-mobile/bind.go:254]: Shouldn't provide focus on only speedway. Should be holistic for all platforms, and keep in mind embeded targets

File: bind/motor-mobile/bind.go (hunk 3 above)
Comment (ntindle) [reply bind/motor-mobile/bind.go:272]: Shouldn't provide focus on only speedway. Should be holistic for all platforms, and keep in mind embeded targets

File: bind/motor-mobile/bind.go (hunk 7 above)
Comment (mcjcloud) [reply bind/motor-mobile/bind.go]: refactored to use constants

File: proto/common/v1/info.proto (hunk 12 above)
Comment (mcjcloud) [reply proto/common/v1/info.proto:48]: Leaving it here

File: bind/motor-mobile/bind.go [hunk 14]
```diff
@@ -214,22 +213,76 @@ func IssuePayment(buf []byte) ([]byte, error) {
...
+	if err := request.Unmarshal(buf); err != nil {
+		return nil, fmt.Errorf("unmarshal request: %s", err)
+	}
+
+	resp, err := instance.BuyAlias(request)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Marshal()
```
Comment (ntindle) [inline bind/motor-mobile/bind.go:251]: This is still needed yes?

File: pkg/crypto/mpc/wallet.go [hunk 15]
```diff
@@ -45,7 +45,6 @@ func GenerateWallet(cb common.MotorCallback, options ...WalletOption) (*Wallet,
 		}(id)
 	}
 	wg.Wait()
-	cb.OnWalletCreated(true)
```
Comment (ntindle) [inline pkg/crypto/mpc/wallet.go:48]: What did this CB do?

File: bind/motor-mobile/bind.go (hunk 14 above)
Comment (mcjcloud) [reply bind/motor-mobile/bind.go:251]: Yes, it now relies on the implementation in pkg/motor

File: pkg/crypto/mpc/wallet.go (hunk 15 above)
Comment (mcjcloud) [reply pkg/crypto/mpc/wallet.go:48]: That must have been old, because that method doesn't exist anymore.

**Review by mcjcloud — APPROVED** [review]

**Review by ntindle — APPROVED** [review]

**Review by joshLong145 — APPROVED** [review]

**Pull request merged**